How To - App Service to Vector DB Auth (#40621)
dotnet/docs: docs/ai/how-to/app-service-db-auth.md

---
title: "Authenticate and Authorize App Service to a Vector Database"
description: "Learn how to authenticate and authorize your App Service .NET application to a vector database solution using Microsoft Entra managed identities, Key Vault, or app settings"
author: haywoodsloan
ms.topic: how-to
ms.date: 04/24/2024
zone_pivot_groups: azure-development-tool-set-two

#customer intent: As a .NET developer, I want authenticate and authorize my App Service to a vector database so that I can securely add memories to the AI in my .NET application.

---

# Authenticate and authorize App Service to a vector database

> This article demonstrates how to manage the connection between your App Service .NET application and a vector database solution. It will cover using Microsoft Entra managed identities for supported services and securely storing connection strings for others.

Reviewer: ```suggestion
This article demonstrates how to manage the connection between your App Service .NET application and a [vector database solution](./conceptual/vector-databases.md). It covers using Microsoft Entra managed identities for supported services and securely storing connection strings for others.
```

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "{myVectorDbReaderRole}" \
--scope "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}"
```

**Subscription scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "{myVectorDbReaderRole}" \
--scope "/subscriptions/{subscriptionId}"
```

**Management group scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "{myVectorDbReaderRole}" \
--scope "/providers/Microsoft.Management/managementGroups/{managementGroupName}"
```

:::zone-end

### Implement token-based authentication with the vector database

The following code samples require these additional libraries:

* [`Azure.Identity` NuGet package](https://www.nuget.org/packages/Azure.Identity)
* [`Microsoft.SemanticKernel.Connectors.AzureAISearch` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel.Connectors.AzureAISearch)

1. Initialize a `DefaultAzureCredential` object to pick up your app's managed identity:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="tokenCredential":::

> 1. Initialize an `IMemoryStore` object for your vector database, then use it to build a `ISemanticTextMemory`:

Reviewer: ```suggestion
1. Initialize an `IMemoryStore` object for your vector database, then use it to build an `ISemanticTextMemory`:
```

## Use Key Vault to store connection secrets

If a vector database doesn't support Microsoft Entra authentication, you can use a Key Vault to store your connection secrets and retrieve them with your App Service application. By using a Key Vault to store your connection secrets you can share them with multiple applications, and control access to individual secrets per application.

Before following these steps, retrieve a connection string for your vector database. For example, see [Use Azure Cache for Redis in .NET Framework](/azure/azure-cache-for-redis/cache-dotnet-how-to-use-azure-redis-cache#retrieve-host-name-ports-and-access-keys-from-the-azure-portal).

### Add a connection string to Key Vault

:::zone target="docs" pivot="azure-portal"

Before following these steps, ensure you have [created a Key Vault using the Azure Portal](/azure/key-vault/general/quick-create-portal).

1. Navigate to your key vault in the [Azure Portal](https://aka.ms/azureportal).
1. In the Key Vault left navigation, select **Objects** then select **Secrets**.
1. Select **+ Generate/Import**.
1. On the **Create a secret** screen choose the following values:

>    * **Upload options**: Manual.

Reviewer: ```suggestion
   * **Upload options**: `Manual`.
```

   * **Name**: Type a name for the secret. The secret name must be unique within a Key Vault.
   * **Value**: The connection string for your vector database
   * Leave the other values to their defaults. Select **Create**.

> 1. Once you receive the message that the secret has been successfully created, it is ready to use in your application.

Reviewer: ```suggestion
1. When you receive the message that the secret has been successfully created, it's ready to use in your application.
```

:::zone-end

:::zone target="docs" pivot="azure-cli"

> Before following these steps, ensure you have [created a Key Vault using the Azure CLI](/azure/key-vault/general/quick-create-cli).

Reviewer: Should this be an alert of some kind?

Author: It's mostly just a pre-req for this specific section, same as with the version of these instruction for the Portal

Author: I think the 'IMPORTANT' alert style would best represent this info. Since it is necessary for success.

1. Grant your user account permissions to your key vault through Role-Based Access Control (RBAC), assign a role using the Azure CLI command `az role assignment create`:

    ```azurecli
    az role assignment create \
    --role "Key Vault Secrets User" \
    --assignee "{yourEmailAddress}" \
    --scope "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{keyVaultName}"
    ```

1. Add the connection string to Key Vault using the Azure CLI command `az keyvault secret set`:

    ```azurecli
    az keyvault secret set \
    --vault-name "{keyVaultName}" \
    --name "{secretName}" \
    --value "{connectionString}"
    ```

:::zone-end

### Grant your App Service access to Key Vault

> 1. Follow these steps to [assign a managed identity to your App Service](#add-a-managed-identity-to-app-service)

Reviewer: ```suggestion
1. [Assign a managed identity to your App Service](#add-a-managed-identity-to-app-service).
```

> 1. Follow these steps to [add the `Key Vault Secrets User` and `Key Vault Reader` roles to your managed identity](#assign-a-azure-role-to-your-managed-identity)

Reviewer: ```suggestion
1. [Add the `Key Vault Secrets User` and `Key Vault Reader` roles to your managed identity](#assign-a-azure-role-to-your-managed-identity).
```

### Implement connection string retrieval from Key Vault

> The following code samples require these additional libraries:

Reviewer: ```suggestion
To use the following code samples, you need these additional libraries :
```

* [`Azure.Identity` NuGet package](https://www.nuget.org/packages/Azure.Identity)
* [`Azure.Extensions.AspNetCore.Configuration.Secrets` NuGet package](https://www.nuget.org/packages/Azure.Extensions.AspNetCore.Configuration.Secrets)
* [`Microsoft.Extensions.Configuration` NuGet package](https://www.nuget.org/packages/Microsoft.Extensions.Configuration)

> These code samples show integrating a Redis database, but they can be applied to any vector database that supports connection strings.

Reviewer: ```suggestion
These code samples use a Redis database, but you can apply them to any vector database that supports connection strings.
```

1. Initialize a `DefaultAzureCredential` object to pick up your app's managed identity:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="tokenCredential":::

1. Add Key Vault when building your configuration, this will map your Key Vault secrets to the `IConfigurationRoot` object:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="vaultConfig":::

> 1. Use the connection string from Key Vault to initialize an `IMemoryStore` object for your vector database, then use it to build a `ISemanticTextMemory`:

Reviewer: ```suggestion
1. Use your vector database connection string from Key Vault to initialize an `IMemoryStore` object, and then use it to build an `ISemanticTextMemory`:
```

```azurecli
az webapp identity assign --name {appName} --resource-group {groupName}
```

:::zone-end

#### Add a user-assigned identity

To create an app with a user-assigned identity, first create the identity and then add its resource identifier to your app config.

:::zone target="docs" pivot="azure-portal"

1. Create a user-assigned managed identity resource according to these [instructions](/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal#create-a-user-assigned-managed-identity).
1. In the left navigation for your app's page, scroll down to the **Settings** group.
1. Select **Identity**.
1. Select **User assigned > Add**.
1. Search for the identity you created earlier, select it, and then select **Add**.

    > [!CAUTION]
    > Once you select **Add**, the app restarts.

:::zone-end

:::zone target="docs" pivot="azure-cli"

1. Create a user-assigned identity:

    ```azurecli
    az identity create --resource-group {groupName} --name {identityName}
    ```

1. Assign the identity to your app:

    ```azurecli
    az webapp identity assign --resource-group {groupName} --name {appName} --identities {identityId}
    ```

:::zone-end

> ### Assign a Azure role to your managed identity

Reviewer 2: ```suggestion
### Assign an Azure role to your managed identity
```

1. Initialize a `DefaultAzureCredential` object to pick up your app's managed identity:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="tokenCredential":::

1. Initialize an `IMemoryStore` object for your vector database, then use it to build an `ISemanticTextMemory`:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="aiStore":::

1. Build a `Kernel` object, then import the `ISemanticTextMemory` object using the `TextMemoryPlugin`:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="addMemory":::

1. Use the `Kernel` object to invoke a prompt that includes memory recall:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="useMemory":::

## Use Key Vault to store connection secrets

If a vector database doesn't support Microsoft Entra authentication, you can use a Key Vault to store your connection secrets and retrieve them with your App Service application. By using a Key Vault to store your connection secrets you can share them with multiple applications, and control access to individual secrets per application.

> Before following these steps, retrieve a connection string for your vector database. For example, see [Use Azure Cache for Redis in .NET Framework](/azure/azure-cache-for-redis/cache-dotnet-how-to-use-azure-redis-cache#retrieve-host-name-ports-and-access-keys-from-the-azure-portal).

Reviewer 2: Can we link to the asp.net core version of a non .NET framework option since that is more legacy?

Maybe here: https://review.learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-web-app-aspnet-core-howto?branch=main

Author: yep, updated link

## Prerequisites

* An Azure account with an active subscription. [Create an account for free](https://azure.microsoft.com/free/?WT.mc_id=A261C142F).
* [.NET SDK](https://dotnet.microsoft.com/download/visual-studio-sdks)
* [`Microsoft.SemanticKernel` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel)
* [`Microsoft.SemanticKernel.Plugins.Memory` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel.Plugins.Memory)
* [Create and deploy a .NET application to App Service](/azure/app-service/quickstart-dotnetcore)
* [Create and deploy a vector database solution](/semantic-kernel/memories/vector-db)

## Use Microsoft Entra managed identity for authentication

If a vector database service supports Microsoft Entra authentication, you can use a managed identity with your App Service to securely access your vector database without having to manually provision or rotate any secrets. For a list of Azure services that support Microsoft Entra authentication, see [Azure services that support Microsoft Entra authentication](/entra/identity/managed-identities-azure-resources/services-id-authentication-support).

> ### Add a managed identity to App Service

Reviewer 3: Is this entire section the same as in the previous article? It might be good to pull it out into a separate include file. I'm just not sure if that works with pivots, but worth a try.

Author: The sections are at different heading levels unfortunately, so the content is nearly the same but not quite.
